crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -3320,6 +3396,101 @@ where
     }
 }
 
+struct BinarySinkNode<C, I1, I2, Op> {
+    id: GlobalNodeId,
+    operator: Op,
+    input_stream1: Stream<C, I1>,
+    input_stream2: Stream<C, I2>,
+    // `true` if both input streams are aliases of the same stream.
+    is_alias: bool,
+}
+
+impl<C, I1, I2, Op> BinarySinkNode<C, I1, I2, Op>
+where
+    I1: Clone,
+    I2: Clone,
+    Op: BinarySinkOperator<I1, I2>,
+    C: Circuit,
+{
+    fn new(
+        operator: Op,
+        input_stream1: Stream<C, I1>,
+        input_stream2: Stream<C, I2>,
+        circuit: C,
+        id: NodeId,
+    ) -> Self {
+        let is_alias = input_stream1.ptr_eq(&input_stream2);
+        Self {
+            id: circuit.global_node_id().child(id),
+            operator,
+            input_stream1,
+            input_stream2,
+            is_alias,
+        }
+    }
+}
+
+impl<C, I1, I2, Op> Node for BinarySinkNode<C, I1, I2, Op>
```

Comment:
does this have to be a sink?
wouldn't it be cleaner to just use an apply2 and a normal sink?

Reply:
That'd work too.  This is a bit cleaner, since we don't need to create a dangling output stream.